starlark/experimental/go.star

```diff
@@ -0,0 +1,128 @@
+# vi:syntax=python
+
+load("github.com/mesosphere/dispatch-catalog/starlark/stable/pipeline@master", "imageResource", "storageResource", "resourceVar")
+load("github.com/mesosphere/dispatch-catalog/starlark/experimental/buildkit@master", "buildkitContainer")
+
+__doc__ = """
+# Go
+
+Provides methods for building and testing Go modules.
+
+Import URL: `github.com/mesosphere/dispatch-catalog/starlark/stable/go`
+"""
+
+def go_test(git, name, paths=None, image="golang:1.13.0-buster", inputs=None, **kwargs):
+    """
+    Run Go tests and generate a coverage report.
+    """
+
+    if not paths:
+        paths = []
+
+    taskName = "{}-test".format(name)
+
+    task(taskName, inputs=[git] + (inputs or []), outputs=[ storageResource(taskName) ], steps=[
+        buildkitContainer(
+            name="go-test-{}".format(name),
+            image=image,
+            command=[ "go", "test", "-v", "-coverprofile", "/workspace/output/{}/coverage.out".format(taskName) ] + paths,
+            env=[ k8s.corev1.EnvVar(name="GO111MODULE", value="on") ],
+            workingDir="/workspace/{}".format(git)
+        ),
+        k8s.corev1.Container(
+            name="coverage-report-{}".format(name),
+            image=image,
+            workingDir="/workspace/{}/".format(git),
+            command=[
+                "sh", "-c",
+                """
+                go tool cover -func /workspace/output/{}/coverage.out | tee /workspace/output/{}/coverage.txt
+                cp /workspace/output/{}/coverage.txt coverage.txt
+                git add coverage.txt
+                git diff --cached coverage.txt
+                """.format(taskName, taskName, taskName)
+            ],
+            env=[ k8s.corev1.EnvVar(name="GO111MODULE", value="on") ],
+        )], **kwargs)
+
+    return taskName
+
+def go(git, name, ldflags=None, os=None, image="golang:1.13.0-buster", inputs=None, **kwargs):
+    """
+    Build a Go binary.
+    """
+
+    if not os:
+        os = ['linux']
```

Comment:
i feel like moving these default values to function params would make it more readable (observed this in other places as well).

Reply:
@takirala in python (and starlark) this is actually correct form (well, technically it should be `if os is None: ...` and not `if not os` (which is subtly different).

The reason is that default values for function arguments are _not_ evaluated from scratch every time: they are like "globals" in that sense:

```python
# This is from my ipython interpreter
In [3]: def foo(name, names=["one"]):
   ...:     names.append(name)
   ...:     print("{}".format(names))
   ...:     

In [4]: foo("gustav")
['one', 'gustav']

In [5]: foo("paul")
['one', 'gustav', 'paul']

```

Reply:
"Principle of least surprise" ... lol

Reply:
Yeah this is why i do that ^

Now, if we're really careful about what the method is doing, it can be okay, but I like to err on the side of caution.

Reply:
Wow, TIL.